### PR TITLE
STBE-10: Teacher registration ep

### PR DESCRIPTION
**Pull request by mohammed-sharraf-sazim**
## Description of Change
- Addition of unique code generation endpoint
- Addition of teacher registration endpoint
## Associated Ticket Link(s)
- https://trello.com/c/8NczZEw8/10-subtask-22-implement-backend-api-for-teacher-registration
- https://trello.com/c/J4S8uUI1/11-subtask-23-implement-error-handling-for-unique-code-verification
## Developer Note(s):
- Github Actions have been removed so 5 changes for that reason
- Made some changes in the teacher entity which lead to a new migration file
- The Teacher module is only created so that I can link the teachers.repository.ts with my teacher entity to follow the nestjs template.
- Main files to look upon are the codes module for generating a unique code for teacher registration and usersService.ts for registering a teacher

### CODE REVIEW

File: src/codes/codes.repository.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      code = Math.floor(100000 + Math.random() * 900000).toString();
+      const existingCode = await this.findByCode(code);
+      if (!existingCode) {
+        isUnique = true;
+      }
+    } while (!isUnique);
+
+    const uniqueCode = new UniqueCode();
+    uniqueCode.code = code;
+    uniqueCode.email = email;
```
Comment (nur-zaman-sazim) [inline src/codes/codes.repository.ts:34]: Please move this to service layer

File: src/codes/codes.service.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+@Injectable()
+export class UniqueCodesService {
+  constructor(
+    private readonly uniqueCodeRepository: UniqueCodeRepository,
+    private readonly entityManager: EntityManager,
+  ) {}
+
+  async generateUniqueCode(email: string): Promise<UniqueCode> {
+    const newGenereatedCode = await this.uniqueCodeRepository.generateUniqueCode(email);
```
Comment (nur-zaman-sazim) [inline src/codes/codes.service.ts:17]: nit: `newGeneratedCode` fix the typo

File: src/codes/codes.service.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    if (uniqueCode.isUsed) {
+      uniqueCode.attempts -= 1;
+      await this.entityManager.persistAndFlush(uniqueCode);
+      throw new BadRequestException(
+        `Unique code already used, you have ${uniqueCode.attempts} attempts remaining`,
+      );
+    }
+
+    return uniqueCode;
+  }
```
Comment (nur-zaman-sazim) [inline src/codes/codes.service.ts:50]:
```suggestion
async validateUniqueCode(code: string, email: string): Promise<UniqueCode | null> {
    const uniqueCode = await this.uniqueCodeRepository.findByEmail(email);

    if (!uniqueCode) {
      throw new BadRequestException(`No code found for the provided email.`);
    }

    if (uniqueCode.isUsed) {
      await this.updateAttempts(uniqueCode);
      throw new BadRequestException(`Unique code has already been used.`);
    }

    if (uniqueCode.code !== code) {
      if (uniqueCode.attempts <= 0) {
        await this.entityManager.removeAndFlush(uniqueCode);
        throw new BadRequestException(`Attempts exceeded. Request a new code.`);
      }
      uniqueCode.attempts -= 1;
      await this.updateAttempts(uniqueCode);
      throw new BadRequestException(`Incorrect code. ${uniqueCode.attempts} attempts remaining.`);
    }

    return uniqueCode;
  }

  private async updateAttempts(uniqueCode: UniqueCode) {
    await this.entityManager.persistAndFlush(uniqueCode);
  }

```

File: src/users/users.controller.ts [hunk 4]
```diff
@@ -9,9 +11,18 @@ export class UsersController {
...
-      newUser,
+      newStudent,
     };
   }
+
+  @Post("register-teacher")
+  async registerTeacher(@Body() createTeacherDto: CreateTeacherDto) {
+    logger.log("register-teacher", createTeacherDto);
+    const newTeacher = await this.usersService.registerTeacher(createTeacherDto);
+    // console.log(newTeacher);
```
Comment (nur-zaman-sazim) [inline src/users/users.controller.ts:24]: please remove comments

**Review by nur-zaman-sazim — CHANGES_REQUESTED** [review]

File: src/codes/codes.repository.ts (hunk 1 above)
Comment (mohammed-sharraf-sazim) [reply src/codes/codes.repository.ts]: Moved it to service layer

File: src/codes/codes.service.ts (hunk 3 above)
Comment (mohammed-sharraf-sazim) [reply src/codes/codes.service.ts]: Suggestion added

File: src/users/users.controller.ts (hunk 4 above)
Comment (mohammed-sharraf-sazim) [reply src/users/users.controller.ts]: Removed

**Review by mahir-mohsin-sazim — APPROVED** [review]

**Pull request merged**